src/metabase/api/notify.clj

```diff
@@ -4,19 +4,37 @@
             [metabase.api.common :as api]
             [metabase.models.database :refer [Database]]
             [metabase.models.table :refer [Table]]
-            [metabase.sync :as sync]))
+            [metabase.sync :as sync]
+            [metabase.sync.sync-metadata :as sync-metadata]
+            [metabase.util.schema :as su]
+            [schema.core :as s]))
+
+(def ^:private ^:dynamic *execute-asynchronously* true)
+
+(defmacro ^:private execute
+  "Emit a runtime check whether to run off thread or on thread. Intended to remove race conditions from tests."
+  [form]
+  `(if *execute-asynchronously* (future ~form) ~form))
```

Comment:
Don't bother changing this, but I have a couple points of feedback on this macro you might want to keep in mind for next time

1. This would be a perfect place to accept an `& body` rather than a single `form`. It costs almost nothing to make it more flexible. Not really necessary here since it's used in only twice, but `& body` is still more idiomatic for macros
2. I've learned to be careful about duplicating forms in macro body, even if we're not worried about multiple evaluation like here because they're in an `if` form. It's good to try to keep the code size down of your macro expansion if you can. Suppose `form` is a hundred lines long -- the macro expansion ends up being 200 lines long. That's inefficient from a code size standpoint, which means it takes longer to compile, takes more space on disk and in memory, and results in more cache misses when ran. Here it makes almost no difference, but I've actually ran into situations where I've got `MethodCodeTooLarge` exceptions during compilation because a macroexpansion generated code that was too big.
3. For one-off tiny macros like this you might want to consider using `clojure.tools.macro/macrolet`
4. Honestly I don't think you really need to make this a macro at all, You could just do something like

   ```clj
   (letfn [(execute [thunk]
             (if *execute-asynchronously*
               (future (thunk))
               (thunk)))]
     ...
     (execute #(table-sync-function table))
     ...)
   ```

Reply:
great points. as i was writing it i was really wondering if i was justifying not writing the lambda myself. I'm normally pretty stingy on that kind of syntax. Especially when getting it wrongs means it runs synchronously in the current thread during a request. Good points and I appreciate you taking the time.